filterExprParser.go

```diff
@@ -0,0 +1,1160 @@
+	var outExpr OrExpr
+	for _, oneExpr := range f.AndConditions {
+		andExpr, err := oneExpr.OutputExpression()
+		if err != nil {
+			return outExpr, err
+		}
+		outExpr = append(outExpr, andExpr)
+	}
+
+	if len(f.SubFilterExpr) > 0 {
+		var combinedExpr AndExpr
+		combinedExpr = append(combinedExpr, outExpr)
+
+		for _, subFilterExpr := range f.SubFilterExpr {
+			subExpr, err := subFilterExpr.OutputExpression()
+			if err != nil {
+				return combinedExpr, err
+			}
+			combinedExpr = append(combinedExpr, subExpr)
+		}
+
+		return combinedExpr, nil
+	} else {
+		return outExpr, nil
+	}
+}
+
+type FEOpenParen struct {
+	Parens string `@"("`
+}
+
+func (feop *FEOpenParen) String() string {
+	return "("
+}
+
+type FECloseParen struct {
+	Parens string `@")"`
+}
+
+func (fecp *FECloseParen) String() string {
+	return ")"
+}
+
+type FEAndCondition struct {
+	OpenParens   []*FEOpenParen  `{ @@ }`
+	OrConditions []*FECondition  `@@ { "AND" @@ }`
+	CloseParens  []*FECloseParen `{ @@ }`
+}
+
+func (ac *FEAndCondition) String() string {
+	output := []string{}
+
+	for _, e := range ac.OpenParens {
+		output = append(output, e.String())
+	}
+
+	first := true
+	for _, e := range ac.OrConditions {
+		if first {
+			first = false
+		} else {
+			output = append(output, "AND")
+		}
+		output = append(output, e.String())
+	}
+
+	for _, e := range ac.CloseParens {
+		output = append(output, e.String())
+	}
+
+	return strings.Join(output, " ")
+}
+
+func (f *FEAndCondition) OutputExpression() (Expression, error) {
+	var outExpr AndExpr
+	for _, oneExpr := range f.OrConditions {
+		expr, err := oneExpr.OutputExpression()
+		if err != nil {
+			return outExpr, err
+		}
+		outExpr = append(outExpr, expr)
+	}
+	return outExpr, nil
+}
+
+type FECondition struct {
+	Not     *FECondition      `"NOT" @@`
+	Operand *FEOperand        `| @@`
+	SubExpr *FilterExpression `| @@`
+}
+
+func (fec *FECondition) String() string {
+	var outputStr []string
+
+	if fec.Not != nil {
+		outputStr = append(outputStr, fmt.Sprintf("NOT %v", fec.Not.String()))
+	} else if fec.Operand != nil {
+		outputStr = append(outputStr, fec.Operand.String())
+	} else if fec.SubExpr != nil {
+		outputStr = append(outputStr, fec.SubExpr.String())
+	} else {
+		outputStr = append(outputStr, "?? (FECondition)")
+	}
+
+	return strings.Join(outputStr, " ")
+}
+
+func (f *FECondition) OutputExpression() (Expression, error) {
+	if f.Not != nil {
+		subNot, err := f.Not.OutputExpression()
+		return NotExpr{subNot}, err
+	} else if f.Operand != nil {
+		return f.Operand.OutputExpression()
+	} else if f.SubExpr != nil {
+		return f.SubExpr.OutputExpression()
+	} else {
+		return nil, fmt.Errorf("Invalid FECondition %v", f.String())
+	}
+}
+
+type FEOperand struct {
+	BooleanExpr *FEBooleanExpr `@@ |`
+	LHS         *FELhs         `( @@ (`
+	Op          *FECompareOp   `( @@`
+	RHS         *FERhs         `@@ ) | `
+	CheckOp     *FECheckOp     `@@ ) )`
+}
+
+func (feo *FEOperand) String() string {
```

Comment:
String() and OutputExpression() are not symmetrical. Maybe they should be. For example, String() does not do LHS != nil check before doing LHS.String().

Reply:
Added checks before outputting